flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationRescaleITCase.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.test.checkpointing;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobgraph.SavepointRestoreSettings;
+import org.apache.flink.runtime.state.AbstractStateBackend;
+import org.apache.flink.runtime.state.StateHandleID;
+import org.apache.flink.streaming.api.environment.CheckpointConfig;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.testutils.junit.SharedReference;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.Preconditions;
+
+import org.junit.Test;
+
+import java.io.File;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.apache.flink.test.util.TestUtils.getMostRecentCompletedCheckpoint;
+
+/**
+ * This verifies that rescale works correctly for Changelog state backend with materialized state /
+ * non-materialized state.
+ */
+public class ChangelogPeriodicMaterializationRescaleITCase
+        extends ChangelogPeriodicMaterializationTestBase {
+
+    public ChangelogPeriodicMaterializationRescaleITCase(
+            AbstractStateBackend delegatedStateBackend) {
+        super(delegatedStateBackend);
+    }
+
+    @Test
+    public void testRescaleOut() throws Exception {
+        testRescale(NUM_SLOTS / 2, NUM_SLOTS);
+    }
+
+    @Test
+    public void testRescaleIn() throws Exception {
+        testRescale(NUM_SLOTS, NUM_SLOTS / 2);
+    }
+
+    private void testRescale(int firstParallelism, int secondParallelism) throws Exception {
+        File firstCheckpointFolder = TEMPORARY_FOLDER.newFolder();
+        JobID firstJobID = generateJobID();
+        SharedReference<Set<StateHandleID>> currentMaterializationId =
+                sharedObjects.add(ConcurrentHashMap.newKeySet());
+
+        StreamExecutionEnvironment env =
+                getEnv(delegatedStateBackend, firstCheckpointFolder, 50, 0, 20, 0);
+        env.getCheckpointConfig()
+                .setExternalizedCheckpointCleanup(
+                        CheckpointConfig.ExternalizedCheckpointCleanup.RETAIN_ON_CANCELLATION);
```

Comment:
I've found that `RETAIN_ON_CANCELLATION` is only applied for the 1st run, before rescaling.
After rescaling, the test fails 1/100 with `FileNotFoundException` while looking for checkpoint metadata.

After adding this config for the 2nd run, the test doesn't fail.

Reply:
Updated. I have also tried to pick your pr for FLINK-26455 to check, and it works.